ffi_spec fails with --release

## Bug Report

Running:
```
make all
make compiler_spec
```
works as expected and spec passes.

Running:
```
make all release=1
make compiler_spec release=1
```
fails with:

```
Failures:

  1) Crystal::FFI::CallInterface .new sum struct
     Failure/Error: pointer_value.should eq 50

       Expected: 50
            got: 0

     # spec/compiler/ffi/ffi_spec.cr:166

Finished in 16:17 minutes
10780 examples, 1 failures, 0 errors, 15 pending

Failed examples:

crystal spec spec/compiler/ffi/ffi_spec.cr:132 # Crystal::FFI::CallInterface .new sum struct
make: *** [Makefile:92: compiler_spec] Error 1
```

Crystal:
```
./bin/crystal -v
Using compiled compiler at .build/crystal
Crystal 1.6.1-dev [809c49e] (2022-10-11)

LLVM: 14.0.6
Default target: x86_64-unknown-linux-gnu
```

This affects the Nix builds: https://github.com/NixOS/nixpkgs/pull/173928
and possibly the solus ones: https://dev.getsol.us/source/crystal/browse/master/package.yml$36 (they remove `spec/compiler/ffi/ffi_spec.cr`)

--- a/spec/compiler/data/ffi/sum.c
+++ b/spec/compiler/data/ffi/sum.c
@@ -1,71 +1,76 @@
 #include <stdarg.h>
+#include <stdint.h>
 #include "../visibility.h"
 
-EXPORT int answer()
+// all the integral return types must be at least as large as the register size
+// to avoid integer promotion by FFI!
+
+EXPORT int64_t answer()
 {
     return 42;
 }
 
-EXPORT int sum(int a, int b, int c)
+EXPORT int64_t sum(int32_t a, int32_t b, int32_t c)
 {
     return a + b + c;
 }
 
 EXPORT void sum_primitive_types(
-    unsigned char a, signed char b,
-    unsigned short c, signed short d,
-    unsigned long e, signed long f,
-    unsigned long long g, signed long long h,
+    uint8_t a, int8_t b,
+    uint16_t c, int16_t d,
+    uint32_t e, int32_t f,
+    uint64_t g, int64_t h,
     float i, double j,
-    long *k)
+    int64_t *k)
 {
-    *k = a + b + c + d + e + f + g + h + (long)i + (long)j + *k;
+    *k = a + b + c + d + e + f + g + h + (int64_t)i + (int64_t)j + *k;
 }
 
 struct test_struct
 {
-    char b;
-    short s;
-    int i;
-    long long j;
+    int8_t b;
+    int16_t s;
+    int32_t i;
+    int64_t j;
     float f;
     double d;
-    int *p;
+    void *p;
 };
 
-EXPORT int sum_struct(struct test_struct s)
+EXPORT int64_t sum_struct(struct test_struct s)
 {
-    *s.p = s.b + s.s + s.i + s.j + s.f + s.d + *(s.p);
-    return *s.p;
+    int64_t *p = (int64_t *)s.p;
+    *p = s.b + s.s + s.i + s.j + s.f + s.d + *p;
+    return *p;
 }
 
-EXPORT int sum_array(int ary[4])
+EXPORT int64_t sum_array(int32_t ary[4])
 {
-    int sum = 0;
-    for (int i = 0; i < 4; i++)
+    int64_t sum = 0;
+    for (int32_t i = 0; i < 4; i++)
     {
         sum += ary[i];
     }
     return sum;
 }
 
-EXPORT int sum_variadic(int count, ...)
+EXPORT int64_t sum_variadic(int32_t count, ...)
 {
     va_list ap;
-    int j;
-    int sum = 0;
+    int32_t j;
+    int64_t sum = 0;
 
     va_start(ap, count); /* Requires the last fixed parameter (to get the address) */
     for (j = 0; j < count; j++)
     {
-        sum += va_arg(ap, int); /* Increments ap to the next argument. */
+        sum += va_arg(ap, int32_t); /* Increments ap to the next argument. */
     }
     va_end(ap);
 
     return sum;
 }
 
-EXPORT struct test_struct make_struct(char b, short s, int i, long long j, float f, double d, void *p)
+EXPORT struct test_struct make_struct(int8_t b, int16_t s, int32_t i, int64_t j, float f, double d, void *p)
 {
     struct test_struct t;
     t.b = b;

--- a/spec/compiler/ffi/ffi_spec.cr
+++ b/spec/compiler/ffi/ffi_spec.cr
@@ -6,6 +6,22 @@ require "compiler/crystal/ffi"
 require "compiler/crystal/loader"
 require "../loader/spec_helper"
 
+# FFI stores a widened return value in `ffi_call` for integral C return types,
+# so we must use an integer type matching the register size
+{% begin %}
+  {% if flag?(:x86_64) %}
+    alias FFISarg = Int64
+  {% elsif flag?(:i386) %}
+    alias FFISarg = Int32
+  {% elsif flag?(:aarch64) %}
+    alias FFISarg = Int64
+  {% elsif flag?(:arm) %}
+    alias FFISarg = Int32
+  {% else %}
+    {% raise "Unsupported target for ffi_sarg" %}
+  {% end %}
+{% end %}
+
 @[Extern]
 private record TestStruct,
   b : LibC::Char,
@@ -33,7 +49,7 @@ describe Crystal::FFI::CallInterface do
       loader = Crystal::Loader.new([SPEC_CRYSTAL_LOADER_LIB_PATH])
       loader.load_library "sum"
       function_pointer = loader.find_symbol("answer")
-      return_value = 0_i32
+      return_value = FFISarg.new(0)
       call_interface.call(function_pointer, Pointer(Pointer(Void)).null, pointerof(return_value).as(Void*))
       return_value.should eq 42
     ensure
@@ -49,7 +65,7 @@ describe Crystal::FFI::CallInterface do
       loader.load_library "sum"
       function_pointer = loader.find_symbol("sum")
 
-      return_value = 0_i32
+      return_value = FFISarg.new(0)
       args = Int32[1, 3, 5]
       arg_pointers = StaticArray(Pointer(Void), 3).new { |i| (args.to_unsafe + i).as(Void*) }
       call_interface.call(function_pointer, arg_pointers.to_unsafe, pointerof(return_value).as(Void*))
@@ -160,7 +176,7 @@ describe Crystal::FFI::CallInterface do
         Pointer(Void).null,
       ]
 
-      return_value = 0i32
+      return_value = FFISarg.new(0)
       call_interface.call(function_pointer, arg_pointers.to_unsafe, pointerof(return_value).as(Void*))
       return_value.should eq 50
       pointer_value.should eq 50
@@ -184,7 +200,7 @@ describe Crystal::FFI::CallInterface do
         loader.load_library "sum"
         function_pointer = loader.find_symbol("sum_array")
 
-        return_value = 0_i32
+        return_value = FFISarg.new(0)
 
         ary = [1, 2, 3, 4]
 
@@ -209,7 +225,7 @@ describe Crystal::FFI::CallInterface do
       loader.load_library "sum"
       function_pointer = loader.find_symbol("sum_variadic")
 
-      return_value = 0_i32
+      return_value = FFISarg.new(0)
       args = Int32[3, 1, 3, 5]
       arg_pointers = StaticArray(Pointer(Void), 4).new { |i| (args.to_unsafe + i).as(Void*) }
       call_interface.call(function_pointer, arg_pointers.to_unsafe, pointerof(return_value).as(Void*))
@@ -225,7 +241,7 @@ describe Crystal::FFI::CallInterface do
       loader.load_library "sum"
       function_pointer = loader.find_symbol("sum_variadic")
 
-      return_value = 1_i32
+      return_value = FFISarg.new(1)
       count = 0
       arg_pointer = pointerof(count).as(Void*)
       call_interface.call(function_pointer, pointerof(arg_pointer), pointerof(return_value).as(Void*))

--- a/spec/compiler/ffi/ffi_spec.cr
+++ b/spec/compiler/ffi/ffi_spec.cr
@@ -8,15 +8,16 @@ require "../loader/spec_helper"
 
 # FFI stores a widened return value in `ffi_call` for integral C return types,
 # so we must use an integer type matching the register size
+
 {% begin %}
   {% if flag?(:x86_64) %}
-    alias FFISarg = Int64
+    alias Crystal::LibFFI::Sarg = Int64
   {% elsif flag?(:i386) %}
-    alias FFISarg = Int32
+    alias Crystal::LibFFI::Sarg = Int32
   {% elsif flag?(:aarch64) %}
-    alias FFISarg = Int64
+    alias Crystal::LibFFI::Sarg = Int64
   {% elsif flag?(:arm) %}
-    alias FFISarg = Int32
+    alias Crystal::LibFFI::Sarg = Int32
   {% else %}
     {% raise "Unsupported target for ffi_sarg" %}
   {% end %}
@@ -49,7 +50,7 @@ describe Crystal::FFI::CallInterface do
       loader = Crystal::Loader.new([SPEC_CRYSTAL_LOADER_LIB_PATH])
       loader.load_library "sum"
       function_pointer = loader.find_symbol("answer")
-      return_value = FFISarg.new(0)
+      return_value = Crystal::LibFFI::Sarg.new(0)
       call_interface.call(function_pointer, Pointer(Pointer(Void)).null, pointerof(return_value).as(Void*))
       return_value.should eq 42
     ensure
@@ -65,7 +66,7 @@ describe Crystal::FFI::CallInterface do
       loader.load_library "sum"
       function_pointer = loader.find_symbol("sum")
 
-      return_value = FFISarg.new(0)
+      return_value = Crystal::LibFFI::Sarg.new(0)
       args = Int32[1, 3, 5]
       arg_pointers = StaticArray(Pointer(Void), 3).new { |i| (args.to_unsafe + i).as(Void*) }
       call_interface.call(function_pointer, arg_pointers.to_unsafe, pointerof(return_value).as(Void*))
@@ -176,7 +177,7 @@ describe Crystal::FFI::CallInterface do
         Pointer(Void).null,
       ]
 
-      return_value = FFISarg.new(0)
+      return_value = Crystal::LibFFI::Sarg.new(0)
       call_interface.call(function_pointer, arg_pointers.to_unsafe, pointerof(return_value).as(Void*))
       return_value.should eq 50
       pointer_value.should eq 50
@@ -200,7 +201,7 @@ describe Crystal::FFI::CallInterface do
         loader.load_library "sum"
         function_pointer = loader.find_symbol("sum_array")
 
-        return_value = FFISarg.new(0)
+        return_value = Crystal::LibFFI::Sarg.new(0)
 
         ary = [1, 2, 3, 4]
 
@@ -225,7 +226,7 @@ describe Crystal::FFI::CallInterface do
       loader.load_library "sum"
       function_pointer = loader.find_symbol("sum_variadic")
 
-      return_value = FFISarg.new(0)
+      return_value = Crystal::LibFFI::Sarg.new(0)
       args = Int32[3, 1, 3, 5]
       arg_pointers = StaticArray(Pointer(Void), 4).new { |i| (args.to_unsafe + i).as(Void*) }
       call_interface.call(function_pointer, arg_pointers.to_unsafe, pointerof(return_value).as(Void*))
@@ -241,7 +242,7 @@ describe Crystal::FFI::CallInterface do
       loader.load_library "sum"
       function_pointer = loader.find_symbol("sum_variadic")
 
-      return_value = FFISarg.new(1)
+      return_value = Crystal::LibFFI::Sarg.new(1)
       count = 0
       arg_pointer = pointerof(count).as(Void*)
       call_interface.call(function_pointer, pointerof(arg_pointer), pointerof(return_value).as(Void*))

--- a/spec/compiler/ffi/ffi_spec.cr
+++ b/spec/compiler/ffi/ffi_spec.cr
@@ -6,31 +6,17 @@ require "compiler/crystal/ffi"
 require "compiler/crystal/loader"
 require "../loader/spec_helper"
 
-# FFI stores a widened return value in `ffi_call` for integral C return types,
-# so we must use an integer type matching the register size
-
-{% begin %}
-  {% if flag?(:x86_64) %}
-    alias Crystal::LibFFI::Sarg = Int64
-  {% elsif flag?(:i386) %}
-    alias Crystal::LibFFI::Sarg = Int32
-  {% elsif flag?(:aarch64) %}
-    alias Crystal::LibFFI::Sarg = Int64
-  {% elsif flag?(:arm) %}
-    alias Crystal::LibFFI::Sarg = Int32
-  {% else %}
-    {% raise "Unsupported target for ffi_sarg" %}
-  {% end %}
-{% end %}
+# all the integral return types must be at least as large as the register size
+# to avoid integer promotion by FFI!
 
 @[Extern]
 private record TestStruct,
-  b : LibC::Char,
-  s : LibC::Short,
-  i : LibC::Int,
-  j : LibC::LongLong,
-  f : LibC::Float,
-  d : LibC::Double,
+  b : Int8,
+  s : Int16,
+  i : Int32,
+  j : Int64,
+  f : Float32,
+  d : Float64,
   p : Pointer(Void)
 
 describe Crystal::FFI::CallInterface do
@@ -45,20 +31,20 @@ describe Crystal::FFI::CallInterface do
 
   describe ".new" do
     it "simple call" do
-      call_interface = Crystal::FFI::CallInterface.new Crystal::FFI::Type.sint16, [] of Crystal::FFI::Type
+      call_interface = Crystal::FFI::CallInterface.new Crystal::FFI::Type.sint64, [] of Crystal::FFI::Type
 
       loader = Crystal::Loader.new([SPEC_CRYSTAL_LOADER_LIB_PATH])
       loader.load_library "sum"
       function_pointer = loader.find_symbol("answer")
-      return_value = Crystal::LibFFI::Sarg.new(0)
+      return_value = 0_i64
       call_interface.call(function_pointer, Pointer(Pointer(Void)).null, pointerof(return_value).as(Void*))
-      return_value.should eq 42
+      return_value.should eq 42_i64
     ensure
       loader.try &.close_all
     end
 
     it "with args" do
-      call_interface = Crystal::FFI::CallInterface.new Crystal::FFI::Type.sint16, [
+      call_interface = Crystal::FFI::CallInterface.new Crystal::FFI::Type.sint64, [
         Crystal::FFI::Type.sint32, Crystal::FFI::Type.sint32, Crystal::FFI::Type.sint32,
       ] of Crystal::FFI::Type
 
@@ -66,11 +52,11 @@ describe Crystal::FFI::CallInterface do
       loader.load_library "sum"
       function_pointer = loader.find_symbol("sum")
 
-      return_value = Crystal::LibFFI::Sarg.new(0)
+      return_value = 0_i64
       args = Int32[1, 3, 5]
       arg_pointers = StaticArray(Pointer(Void), 3).new { |i| (args.to_unsafe + i).as(Void*) }
       call_interface.call(function_pointer, arg_pointers.to_unsafe, pointerof(return_value).as(Void*))
-      return_value.should eq 9
+      return_value.should eq 9_i64
     ensure
       loader.try &.close_all
     end
@@ -89,7 +75,7 @@ describe Crystal::FFI::CallInterface do
       loader.load_library "sum"
       function_pointer = loader.find_symbol("sum_primitive_types")
 
-      pointer_value = 11_i32
+      pointer_value = 11_i64
       arg_pointers = StaticArray[
         Pointer(UInt8).malloc(1, 1).as(Void*),
         Pointer(Int8).malloc(1, 2).as(Void*),
@@ -101,7 +87,7 @@ describe Crystal::FFI::CallInterface do
         Pointer(Int64).malloc(1, 8).as(Void*),
         Pointer(Float32).malloc(1, 9.0).as(Void*),
         Pointer(Float64).malloc(1, 10.0).as(Void*),
-        Pointer(Int32*).malloc(1, pointerof(pointer_value)).as(Void*),
+        Pointer(Int64*).malloc(1, pointerof(pointer_value)).as(Void*),
         Pointer(Void).null,
       ]
 
@@ -147,7 +133,7 @@ describe Crystal::FFI::CallInterface do
     end
 
     it "sum struct" do
-      call_interface = Crystal::FFI::CallInterface.new Crystal::FFI::Type.sint32, [
+      call_interface = Crystal::FFI::CallInterface.new Crystal::FFI::Type.sint64, [
         Crystal::FFI::Type.struct([
           Crystal::FFI::Type.sint8,
           Crystal::FFI::Type.sint16,
@@ -163,7 +149,7 @@ describe Crystal::FFI::CallInterface do
       loader.load_library "sum"
       function_pointer = loader.find_symbol("sum_struct")
 
-      pointer_value = 11_i32
+      pointer_value = 11_i64
       arg_pointers = StaticArray[
         Pointer(TestStruct).malloc(1, TestStruct.new(
           b: 2,
@@ -177,10 +163,10 @@ describe Crystal::FFI::CallInterface do
         Pointer(Void).null,
       ]
 
-      return_value = Crystal::LibFFI::Sarg.new(0)
+      return_value = 0_i64
       call_interface.call(function_pointer, arg_pointers.to_unsafe, pointerof(return_value).as(Void*))
-      return_value.should eq 50
-      pointer_value.should eq 50
+      return_value.should eq 50_i64
+      pointer_value.should eq 50_i64
     ensure
       loader.try &.close_all
     end
@@ -188,7 +174,7 @@ describe Crystal::FFI::CallInterface do
     # passing C array by value is not supported everywhere
     {% unless flag?(:win32) %}
       it "array" do
-        call_interface = Crystal::FFI::CallInterface.new Crystal::FFI::Type.sint32, [
+        call_interface = Crystal::FFI::CallInterface.new Crystal::FFI::Type.sint64, [
           Crystal::FFI::Type.struct([
             Crystal::FFI::Type.sint32,
             Crystal::FFI::Type.sint32,
@@ -201,7 +187,7 @@ describe Crystal::FFI::CallInterface do
         loader.load_library "sum"
         function_pointer = loader.find_symbol("sum_array")
 
-        return_value = Crystal::LibFFI::Sarg.new(0)
+        return_value = 0_i64
 
         ary = [1, 2, 3, 4]
 
@@ -211,7 +197,7 @@ describe Crystal::FFI::CallInterface do
         ]
 
         call_interface.call(function_pointer, arg_pointers.to_unsafe, pointerof(return_value).as(Void*))
-        return_value.should eq 10
+        return_value.should eq 10_i64
       ensure
         loader.try &.close_all
       end
@@ -220,43 +206,43 @@ describe Crystal::FFI::CallInterface do
 
   describe ".variadic" do
     it "basic" do
-      call_interface = Crystal::FFI::CallInterface.variadic Crystal::FFI::Type.sint16, [Crystal::FFI::Type.sint32, Crystal::FFI::Type.sint32, Crystal::FFI::Type.sint32, Crystal::FFI::Type.sint32] of Crystal::FFI::Type, 1
+      call_interface = Crystal::FFI::CallInterface.variadic Crystal::FFI::Type.sint64, [Crystal::FFI::Type.sint32, Crystal::FFI::Type.sint32, Crystal::FFI::Type.sint32, Crystal::FFI::Type.sint32] of Crystal::FFI::Type, 1
 
       loader = Crystal::Loader.new([SPEC_CRYSTAL_LOADER_LIB_PATH])
       loader.load_library "sum"
       function_pointer = loader.find_symbol("sum_variadic")
 
-      return_value = Crystal::LibFFI::Sarg.new(0)
+      return_value = 0_i64
       args = Int32[3, 1, 3, 5]
       arg_pointers = StaticArray(Pointer(Void), 4).new { |i| (args.to_unsafe + i).as(Void*) }
       call_interface.call(function_pointer, arg_pointers.to_unsafe, pointerof(return_value).as(Void*))
-      return_value.should eq 9
+      return_value.should eq 9_i64
     ensure
       loader.try &.close_all
     end
 
     it "zero varargs" do
-      call_interface = Crystal::FFI::CallInterface.variadic Crystal::FFI::Type.sint16, [Crystal::FFI::Type.sint32] of Crystal::FFI::Type, 1
+      call_interface = Crystal::FFI::CallInterface.variadic Crystal::FFI::Type.sint64, [Crystal::FFI::Type.sint32] of Crystal::FFI::Type, 1
 
       loader = Crystal::Loader.new([SPEC_CRYSTAL_LOADER_LIB_PATH])
       loader.load_library "sum"
       function_pointer = loader.find_symbol("sum_variadic")
 
-      return_value = Crystal::LibFFI::Sarg.new(1)
-      count = 0
+      return_value = 1_i64
+      count = 0_i32
       arg_pointer = pointerof(count).as(Void*)
       call_interface.call(function_pointer, pointerof(arg_pointer), pointerof(return_value).as(Void*))
-      return_value.should eq 0
+      return_value.should eq 0_i64
     ensure
       loader.try &.close_all
     end
 
     it "validates args size" do
       expect_raises Exception, "invalid value for fixed_args" do
-        Crystal::FFI::CallInterface.variadic Crystal::FFI::Type.sint32, [] of Crystal::FFI::Type, 1
+        Crystal::FFI::CallInterface.variadic Crystal::FFI::Type.sint64, [] of Crystal::FFI::Type, 1
       end
       expect_raises Exception, "invalid value for fixed_args" do
-        Crystal::FFI::CallInterface.variadic Crystal::FFI::Type.sint32, [] of Crystal::FFI::Type, -1
+        Crystal::FFI::CallInterface.variadic Crystal::FFI::Type.sint64, [] of Crystal::FFI::Type, -1
       end
     end
   end

